---
title: "GAGA phylogeny with HGT highlights"
output:
  html_document:
    toc: true
    toc_float: true
    toc_depth: 2
    collapsed: true
    number_sections: true
    df_print: paged
  pdf_document: default
---
<STYLE TYPE="text/css">
<!--
  td{
    font-family: Arial;
    font-size: 8pt;
    padding:0px;
    cellpadding="0";
    cellspacing="0"
  }
  th {
    font-family: Arial;
    font-size: 8pt;
    height: 20px;
    font-weight: bold;
    text-align: right;
    background-color: #ccccff;
  }
  table {
    border-spacing: 0px;
    border-collapse: collapse;
  }
--->
</STYLE>

```{r global-options, include=FALSE}
# Define the global chunk options
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, message=FALSE)
```
Load the necessary R libraries
```{r libraries}
#if (!requireNamespace("BiocManager", quietly = TRUE))
   #install.packages("BiocManager")

#BiocManager::install("treeio")
#install.packages("devtools")
#devtools::install_github("GuangchuangYu/ggtree")
library(treeio)
library(ggtree)
library(phytools)
library(viridisLite)
library(ggplot2)
library(googlesheets4)
library(dplyr)
require(ape)
library(tidytree)
```

Load data from google spreadsheet (Already stored in an R object that you can just load)
```{r}
#id<-"1e7Hj_BQ8rke5XW7j93YL6QVAGbcvsb2yw7oD86WnPJY"
#gsheet<-read_sheet(id,sheet="GlobalOverview",skip = 2,na = c(""," ","NA","#DIV/0!"),col_types = "c",trim_ws=T)
#GlobalOverview<-as.data.frame(gsheet,stringsAsFactors=F)
#save(GlobalOverview,file="../../misc/data/gsheet.GlobalOverview.21-09-22.Robj")
load("/Users/Janina/sciebo/MASTER/Masterarbeit/lgt/0_data/gsheet.GlobalOverview.21-09-22.Robj")

load("/Users/Janina/sciebo/MASTER/Masterarbeit/lgt/0_data/GAGA.tree.plot4.Robject")
```

Create a dataframe where the first row is the GAGA id (e.g. GAGA-0002), which are also used as the tip labels in the phylogeny
```{r}
data2add<-data.frame(GAGA.id = GlobalOverview$`GAGA ID`, GlobalOverview)
```

Gather subfamily info for NCBI IDs as well and manually add the subfamily
```{r}
colnames(data2add)

subset(data2add,is.na(Subfamily))

data2add$Subfamily[data2add$GAGA.id=="NCBI-0001"]<-"Dorylinae"
data2add$Subfamily[data2add$GAGA.id=="NCBI-0002"]<-"Myrmicinae"
data2add$Subfamily[data2add$GAGA.id=="NCBI-0003"]<-"Myrmicinae"
data2add$Subfamily[data2add$GAGA.id=="NCBI-0004"]<-"Myrmicinae"
data2add$Subfamily[data2add$GAGA.id=="NCBI-0005"]<-"Formicinae"
data2add$Subfamily[data2add$GAGA.id=="NCBI-0006"]<-"Myrmicinae"
data2add$Subfamily[data2add$GAGA.id=="NCBI-0007"]<-"Ponerinae"
data2add$Subfamily[data2add$GAGA.id=="NCBI-0008"]<-"Formicinae"
data2add$Subfamily[data2add$GAGA.id=="NCBI-0009"]<-"Ponerinae"
data2add$Subfamily[data2add$GAGA.id=="NCBI-0010"]<-"Dolichoderinae"
data2add$Subfamily[data2add$GAGA.id=="NCBI-0011"]<-"Ponerinae"
data2add$Subfamily[data2add$GAGA.id=="NCBI-0012"]<-"Myrmicinae"
data2add$Subfamily[data2add$GAGA.id=="NCBI-0013"]<-"Myrmicinae"
data2add$Subfamily[data2add$GAGA.id=="NCBI-0014"]<-"Myrmicinae"
data2add$Subfamily[data2add$GAGA.id=="NCBI-0015"]<-"Myrmicinae"
data2add$Subfamily[data2add$GAGA.id=="NCBI-0016"]<-"Myrmicinae"
data2add$Subfamily[data2add$GAGA.id=="NCBI-0017"]<-"Myrmicinae"
data2add$Subfamily[data2add$GAGA.id=="OUT-0001"]<-"Formicinae"
data2add$Subfamily[data2add$GAGA.id=="OUT-0002"]<-"Formicinae"
```

Create a column with clean species IDs for all species.

This has to be done, because the database of all the sequenced species is a bit convoluted.
The function coalesce() finds the first non-missing value at each position. In this case, the columns "Species" and "Species..from.Assembly.Overview" are collapsed. If the column "Species" does not have an entry, the entry from the column "Species..from.Assembly.Overview" will be taken.
```{r}
data2add$cleanSpeciesID<-dplyr::coalesce(data2add$Species,data2add$Species..from.Assembly.Overview.)
```


Load the tree file of all the GAGA genomes.
```{r}
#tree <- read.iqtree("/Users/Janina/sciebo/MASTER/Masterarbeit/lgt/0_data/hymenoptera_psc98_partition_b1000_40threads_msubnucl.treefile")

tree<-read.tree("/Users/Janina/sciebo/MASTER/Masterarbeit/lgt/0_data/GAGA_dated_phylogeny_newick.tre")
```

Create a very quick and dirty plot of the phylogeny
```{r}
plot(tree@phylo,show.node.label = T,show.tip.label = T,cex=.5)
nodelabels(bg=rgb(0,0,0,0),col="red",frame = "none")
```

Check the plot to see where the root should be. Here we root the tree at node 166 that separates the outgroups, the honeybee "Amel" (Apis mellifera) and the parasitoid wasp Nasonia (Nvit), from ants.
```{r}
#tree.rooted<-treeio::root(tree,node=166)
tree.rooted <- tree
tree.rooted.subset<-treeio::drop.tip(tree.rooted,tip=c("NCBI-0018_Amel","NCBI-0019_Nvit"))
```

Plot basic tree
```{r}
tp<-ggtree(tree.rooted.subset,ladderize = T, layout = "circular")+geom_rootedge(.01)
```

Add the species identity from `FAS.genes.and.names.tsv`
```{r}
tp$data$label[tp$data$isTip==T]<-gsub("_.*","",tp$data$label[tp$data$isTip==T])

tp2<-tp %<+% data2add
tp2 
```

# A) Make complex tree with the subfamilies highlighted by color
```{r}
#install.packages("ggnewscale")
library("ggnewscale")

#Make Subfamily as a factor (discrete color scale for factors)
tp2$data$Subfamily<-as.factor(tp2$data$Subfamily)

tp3<-tp2 +
  geom_tiplab2(aes(label=paste(label,cleanSpeciesID,sep=" "), color = Subfamily),size=2,align = T,linesize = .1)+    #add the tip labels
  geom_nodepoint(aes(fill=UFboot),stroke=0.5,pch=21,size=2)+       #add the info about bootstrapping success as points of different colors to the nodes
 scale_fill_viridis_c()+  #change the color of the UFboot node points to a viridis color scale
  theme(legend.position = "left",
        legend.title=element_text(size=11), # the title size of legend.
        legend.text=element_text(size=10)
        )

# tp2.5 is a less complicated tree without UFboot
tp2.5<-tp2 +
  geom_tiplab2(aes(label=paste(cleanSpeciesID,sep=" "), color = Subfamily),size=2,align = T,linesize = .1) # remove GAGAid from label 
```

```{r}
tp3 #plot the tree
#ggsave("/Users/Janina/Documents/MASTER/Masterarbeit/plots/GAGA_phylogenies/Subfamilies.pdf",
#width = 15,height=30) #save the tree in a pdf

tp2.5
```


# A.1) LGT data: Include the LGT data in the circular phylogeny
```{r}
library(readxl)

LGTs<-read_excel("/Users/Janina/sciebo/MASTER/Masterarbeit/lgt/0_data/LGTs.LocusLevel.UniProt.Info.Reannotated.xlsx")
```


# A.2) LGT data: Add the LGT candidate numbers per genome to the phylogeny
Plot the circular phylogeny and add red dot to the tip, where the size of the red dot represents the number of identified LGTs.
```{r}
#Create a tally dataframe that counts how many LGT candidates we have per species.
LGT.tally<-as.data.frame(table(select(LGTs,.id)))
colnames(LGT.tally)<-c("GAGAid","LGT.count")

#Add the LGT.tally dataframe to the phylogeny object as additional data.
tp4<-tp3 %<+% LGT.tally
#see tp4$data$LGT.count

#Plot the phylogeny and add red dot to the tip, where the size of the red dot represents the number of identified LGTs
tp5<-tp4+geom_tippoint(aes(size=LGT.count),col="red", alpha=.5)
tp5
#ggsave("/Users/Janina/Documents/MASTER/Masterarbeit/plots/GAGA_phylogenies/GAGA.phylogeny.LGTs.pdf",width = 12,height=24) #save the tree in a pdf

#write.table(LGT.tally, file="GAGAspecies_with_lgts.tsv", sep=",", col.names = NA)
```


```{r}
#BiocManager::install("ggstar")
#BiocManager::install("ggtreeExtra")
library(ggtree)
library(ggtreeExtra)
library(ggplot2)
library(ggnewscale)
library(dplyr)
library(tidytree)
library(ggstar)
```

# A.3) Plot all predicted proteins onto the phylogeny
```{r}
colnames(LGTs)

# create grouping for protein type
LGTs$type<-"other"
LGTs$type[grep("Ank",LGTs$Protein.names)]<-"Ank"
LGTs$type[grep("Cyclopropane|EC 2.1.1.79",LGTs$Protein.names)]<-"CFA"
LGTs$type[grep("N-acetylmuramic",LGTs$Protein.names)]<-"MurNAc"
LGTs$type[grep("DUF1016|Lysozyme|lysozyme",LGTs$Protein.names)]<-"Lys"

# count the number of proteins per species
LGT.proteins<-select(LGTs,c("type",".id"))
LGT.proteins <- LGT.proteins %>% count(.id,type, sort = T)
colnames(LGT.proteins)<-c("GAGAid","protein","count")

LGT.proteins <- na.omit(LGT.proteins)
head(LGT.proteins)
```

```{r}
library("viridis")
# For the protein bar layer outside of the tree
tp5.5 <- tp2.5 +
  new_scale_fill() +
  geom_fruit(
          data=LGT.proteins,
          geom=geom_bar,
          mapping=aes(y=GAGAid, x=count,fill=protein), #The count will be mapped to x
          pwidth=.4,      # Width of the external layer
          stat="identity",
          orientation="y", #The orientation of the axis
          position=position_stackx(hexpand=.45),   #Adjust the position of the geom_layer
          axis.params =list(
                         axis="x",       # add axis text of the layer.
                         vjust=0.2,
                         text.angle=0,   # the text size of axis.
                         text.size=1.8,
                         hjust=1,        # adjust the horizontal position of text of axis
                         nbreak=1,       # number of lines within geom_layer,
                           ),
         grid.params=list(alpha=.2)      # Parameter to adjust gridlines around the tree
      ) +
      scale_fill_discrete(     #color of the external geom_layer
       name="HGT-encoded protein",        #name of the legend for the external geom_layer
       guide=guide_legend(keywidth=1,keyheight=1,order=6, ncol=6)
      ) +
      theme(legend.position="bottom", # the position of legend.
            legend.box="vertical", legend.margin = margin(0,0,0,0), #change the spacing between the legend and the plot
            legend.box.margin=margin(10,10,10,10),
          legend.background=element_rect(fill=NA), # the background of legend.
          legend.title=element_text(size=12), # the title size of legend.
          legend.text=element_text(size=10) # the text size of legend.
          #legend.spacing.y = unit(0.01, "cm")
      )
tp5.5
```

# A.5) Make a dataframe for the bacteria and subset the LGT count into Wolbachia and Blochmannia origin
```{r}
LGT.bacteria<-read_excel("/Users/Janina/sciebo/MASTER/Masterarbeit/lgt/0_data/LGTs.LocusLevel.Expression.xlsx")[,c(15,2)]

# Make categories for all bacteria
LGT.bacteria$bac.type<-"other"
LGT.bacteria$bac.type[grep("Blochmannia|Baumannia|Yersinia|Sodalis|Hamilton|Cardinium|Buchnera|Westeberhardia|Pantoea|Shigella",LGT.bacteria$prokaryote_origin)]<-"Blochmannia-like"
LGT.bacteria$bac.type[grep("Spiroplasma",LGT.bacteria$prokaryote_origin)]<-"Spiroplasma"
LGT.bacteria$bac.type[grep("Wolbachia",LGT.bacteria$prokaryote_origin)]<-"Wolbachia"

# Count the number of each different bacterium per species
LGT.bac <- LGT.bacteria %>% count(.id,bac.type, sort = T)

# Count the total number of all bacteria per species
LGT.bac2 <- LGT.bacteria %>% count(.id, sort = T)
LGT.bac <- na.omit(LGT.bac)
LGT.bac2 <- na.omit(LGT.bac2)

# Merge both datasets to plot the bacteria as relative count
LGT.bac<-merge(LGT.bac,LGT.bac2,by=".id",all.x=T)
colnames(LGT.bac)<-c("GAGAid","prokaryote_origin","count","totalCount")

# Calculate the relative bacterial count of e.g Wolbachia (for example 80 % Wolbachia, 20 % Blochmannia)
LGT.bac$relative<-LGT.bac$count/LGT.bac$totalCount
```

# A.6) Add the bacterial origin
```{r}
tp6 <- tp5.5 +
  new_scale_fill() +
  geom_fruit(
          data=LGT.bac,
          geom=geom_bar,
          mapping=aes(y=GAGAid, x=relative,fill=prokaryote_origin), #The count will be mapped to x
          pwidth=.1,      # Width of the external layer
          stat="identity",
          orientation="y", #The orientation of the axis
          position=position_stackx(hexpand=.4),   #Adjust the position of the geom_layer
          axis.params =list(
                          axis="x",       # add axis text of the layer.
                          vjust=0.2,
                          text.angle=0,   # the text size of axis.
                          text.size=1.5,
                          hjust=1,        # adjust the horizontal position of text of axis
                          nbreak=1,       # number of lines within geom_layer,
                          ),
          grid.params=list(alpha=.2)      # Parameter to adjust gridlines around the tree
      ) +
      scale_fill_brewer(    #color of the external geom_layer
      
       name="HGT Prokaryote origin",        #name of the legend for the external geom_layer
       guide=guide_legend(keywidth=1,keyheight=1,order=6, ncol=6)
      ) +
      theme(legend.position="bottom", # the position of legend.
            legend.box="vertical", legend.margin = margin(0,0,0,0), #change the spacing between the legend and the plot
            legend.box.margin=margin(20,20,20,20), #change the spacing between the legend and the plot
          legend.background=element_rect(fill=NA), # the background of legend.
          legend.title=element_text(size=12), # the title size of legend.
          legend.text=element_text(size=10) # the text size of legend.
         # legend.spacing.y = unit(0.01, "cm")
      )
tp6

#ggsave("/Users/Janina/Documents/MASTER/Masterarbeit/plots/GAGA_phylogenies/GAGA.phylogeny.proteins.bacteria_RAW.pdf", width=20, height=25)
#ggsave("/Users/Janina/Documents/MASTER/Masterarbeit/plots/GAGA_phylogenies/try.pdf",width=13,height=26) #save the tree in a pdf
```

# A.7) Separating the legend from the tree to position it below the tree
```{r}
tp_bacteria <- tp6
tp_protein <- tp5.5

require(cowplot)
leg1 <- get_legend(tp_bacteria)

tp7 <- tp6 + theme(legend.position="none")
plot_grid(tp7, leg1, ncol=1, rel_widths=c(0.07, 0.5))
```

```{r}
tp7 <- tp6 +
      theme(legend.position="bottom", # the position of legend.
          legend.background=element_rect(), # the background of legend.
          legend.title=element_text(size=11), # the title size of legend.
          legend.text=element_text(size=9), # the text size of legend.
          legend.spacing.y = unit(0.02, "cm")
      )
tp7

#ggsave("/Users/Janina/Documents/MASTER/Masterarbeit/plots/GAGA_phylogenies/GAGA.phylogeny.proteins.bacterialorigin.pdf",width=15,height=30) #save the tree in a pdf
#graph2ppt(file="/Users/Janina/Documents/MASTER/Masterarbeit/plots/GAGA_phylogenies/GAGA.phylogeny.proteins.bacterialorigin.ppt", width=12, height=12)
```