---
title: "GAGA phylogeny with HGT highlights"
output:
  html_document:
    toc: true
    toc_float: true
    toc_depth: 2
    collapsed: true
    number_sections: true
    df_print: paged
  pdf_document: default
---
<STYLE TYPE="text/css">
<!--
  td{
    font-family: Arial;
    font-size: 8pt;
    padding:0px;
    cellpadding="0";
    cellspacing="0"
  }
  th {
    font-family: Arial;
    font-size: 8pt;
    height: 20px;
    font-weight: bold;
    text-align: right;
    background-color: #ccccff;
  }
  table {
    border-spacing: 0px;
    border-collapse: collapse;
  }
--->
</STYLE>

```{r global-options, include=FALSE}
# Define the global chunk options
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, message=FALSE)
```

Load the necessary R libraries
```{r libraries}
#if (!requireNamespace("BiocManager", quietly = TRUE))
   #install.packages("BiocManager")

#BiocManager::install("treeio")
#install.packages("devtools")
#devtools::install_github("GuangchuangYu/ggtree")

#BiocManager::install("ggstar")
#BiocManager::install("ggtreeExtra")
library(treeio)
library(ggtree)
library(phytools)
library(viridisLite)
library(ggplot2)
library(googlesheets4)
library(dplyr)
require(ape)
library(tidytree)
library(ggnewscale)
library(ggtreeExtra)
library(ggnewscale)
library(reshape2)
library(ggstar)
library(TDbook)
```

## Load data 
(GAGA_dated_phylogeny_newick_woutgroup_DatesPeters2017_CurrBio.tre)
```{r}
# Load the id2species table
GAGA.data<-read.table("./0_data/Tables/all.GAGA.data.tsv",sep="\t",T)

# Load tree itself
GAGA2.tree<-read.tree("./0_data/trees/GAGA_dated_phylogeny_newick_woutgroup_DatesPeters2017_CurrBio.tre")
```

## Subset tree with treeio
```{r}
# Remove Outgroup tips from the tree
tips_to_remove<-c("OUTG-0008", "OUTG-0007", "OUTG-0006", "OUTG-0005", "OUTG-0004", "OUTG-0003","OUTG-0002", "OUTG-0001")

subset_tree<-drop.tip(GAGA2.tree, tips_to_remove)

# Remove Outgroups from GAGA.data
GAGA2.data<-GAGA.data %>%
  dplyr::filter(!grepl("OUTG", GAGA.data$label))

# Rename the label column to GAGAid to avoid confusion
GAGA2.data<-rename(GAGA2.data, GAGAid = label)
```


# Define variables for tree plotting
```{r}
offset1<-48 #Define how far away the species names should be from the main tree
hex1<-178 # presence/absence
hex2<-160 # donor
```

## Plot tree
```{r}
# Add species labels to tree and replace GAGAids
df <- data.frame(label = GAGA2.data$GAGAid, 
                Species = GAGA2.data$DisplayName,
                Subfamily = GAGA2.data$Subfamily,
                node=GAGA2.data$node)

# Nodes need to be replaced accordingly
df$node <- seq(1, 163)

subset_tree2 <- full_join(subset_tree, df, by = "label")
subset_tree2

p0<-ggtree(subset_tree, layout='circular')
p1 <- p0 + geom_tiplab()
p1

p2<-p0 %<+% df +
  geom_tiplab(aes(label=Species),
  offset=offset1,
  size=3) +
  theme(legend.position="bottom")
p2

# Plot node labels to color subfamilies accordingly
p2 + geom_text(aes(label=node))


# Define nodes to color subfamilies based on
nodes <- data.frame(node=c(250,1, 167, 188,192, 30, 195, 200, 206, 207, 210, 249), type = c("Myrmicinae", "Leptanillinae", "Ponerinae", "Proceratiinae", "Amblyoponinae", "Paraponerinae", "Dorylinae", "Dolichoderinae", "Pseudomyrmecinae", "Myrmeciinae", "Formicinae", "Ectatomminae"))

# Highlight subfamilies by color
p3 <- p2 +
  geom_hilight(data=nodes, aes(node=node, fill=type), type = "gradient", alpha = .3) +
  scale_color_gradient()
p3
```

# A.1) LGT data: Include the LGT data in the circular phylogeny
```{r}
library(readxl)
library(openxlsx)

LGTs<-read_excel("./0_data/Tables/LGTs.LocusLevel.UniProt.Info.Reannotated.xlsx")
```


# A.2) LGT data: Add the LGT candidate numbers per genome to the phylogeny
Plot the circular phylogeny and add red dot to the tip, where the size of the red dot represents the number of identified LGTs.
```{r}
#Create a tally dataframe that counts how many LGT candidates we have per species.
LGT.tally<-as.data.frame(table(select(LGTs,.id)))
colnames(LGT.tally)<-c("GAGAid","LGT.count")

#Add the LGT.tally dataframe to the phylogeny object as additional data.
p5<-p3 %<+% LGT.tally

#Plot the phylogeny and add red dot to the tip, where the size of the red dot represents the number of identified LGTs
p6<-p5+geom_tippoint(aes(size=LGT.count),col="red", alpha=.5)
p6
#ggsave(p6,file="./plots/GAGA_dated_phylogenies/GAGA.phylogeny.LGTs.dots.pdf",width = 12,height=24) #save the tree in a pdf

#write.table(LGT.tally, file="GAGAspecies_with_lgts.tsv", sep=",", col.names = NA)
```


# A.3) Plot all predicted proteins onto the phylogeny
```{r}
colnames(LGTs)

# create grouping for protein type
LGTs$type<-"other"
LGTs$type[grep("Ank",LGTs$Protein.names)]<-"Ank"
LGTs$type[grep("Cyclopropane|EC 2.1.1.79",LGTs$Protein.names)]<-"CFA"
LGTs$type[grep("N-acetylmuramic",LGTs$Protein.names)]<-"MurNAc"
LGTs$type[grep("DUF1016|Lysozyme|lysozyme",LGTs$Protein.names)]<-"Lys"
LGTs$type[grep("Ribosomal RNA",LGTs$Protein.names)]<-"metA"

# count the number of proteins per species
LGT.proteins<-select(LGTs,c("type",".id"))
LGT.proteins <- LGT.proteins %>% count(.id,type, sort = T)
colnames(LGT.proteins)<-c("GAGAid","protein","count")

LGT.proteins <- na.omit(LGT.proteins)
head(LGT.proteins)
```

```{r}
library("viridis")

#switch to a simple presence/absence visualization
LGT.proteins$count[LGT.proteins$count>1]<-1
# For the protein bar layer outside of the tree
tp1 <- p3 +
  new_scale_fill() +
  geom_fruit(
          data=LGT.proteins,
          geom=geom_bar,
          mapping=aes(y=GAGAid, x=count,fill=protein), #The count will be mapped to x
          pwidth=0.2,      # Width of the external layer
          stat="identity",
          orientation="y", #The orientation of the axis
          position=position_stackx(hexpand=hex1),   #Adjust the position of the geom_layer
          axis.params =list(
                         axis="x",       # add axis text of the layer.
                         vjust=0.2,
                         text.angle=0,   # the text size of axis.
                         text.size=1.8,
                         hjust=1.5,        # adjust the horizontal position of text of axis
                         nbreak=1,       # number of lines within geom_layer,
                           ),
         grid.params=list(alpha=.2)      # Parameter to adjust gridlines around the tree
      ) +
    #  scale_fill_discrete(     #color of the external geom_layer
    #   name="HGT-encoded protein",        #name of the legend for the external geom_layer
    #   guide=guide_legend(keywidth=1,keyheight=1,order=6, ncol=6)
    #  ) +
      theme(legend.position="bottom", # the position of legend.
            legend.box="vertical", legend.margin = margin(0,0,0,0), #change the spacing between the legend and the plot
            legend.box.margin=margin(10,10,10,10),
          legend.background=element_rect(fill=NA), # the background of legend.
          legend.title=element_text(size=12), # the title size of legend.
          legend.text=element_text(size=10) # the text size of legend.
          #legend.spacing.y = unit(0.01, "cm")
      )
tp1<-tp1+scale_fill_manual(values=c("steelblue","red","#EF7F9A","firebrick", "darkorange","steelblue2"))

tp1
```

# A.5) Make a dataframe for the bacteria and subset the LGT count into Wolbachia and Blochmannia origin
```{r}
LGT.bacteria<-read_excel("./0_data/Tables/LGTs.LocusLevel.Expression.xlsx")[,c(1,2,17,16,15,12,13)]

# Make categories for all bacteria
LGT.bacteria$bac.type<-"other"
LGT.bacteria$bac.type[grep("Enterobacterales",LGT.bacteria$order_prokaryote_origin)]<-"Blochmannia-like (Gammaproteobacteria)"
LGT.bacteria$bac.type[grep("Spiroplasma|Mycoplasma",LGT.bacteria$prokaryote_origin)]<-"Spiroplasma/Mycoplasma (Mollicutes)"
LGT.bacteria$bac.type[grep("Cardinium",LGT.bacteria$prokaryote_origin)]<-"Cardinium (Sphingobacteriia)"
LGT.bacteria$bac.type[grep("Wolbachia",LGT.bacteria$prokaryote_origin)]<-"Wolbachia (Alphaproteobacteria)"

# Count the number of each different bacterium per species
LGT.bac <- LGT.bacteria %>% count(.id,bac.type, sort = T)

# Count the total number of all bacteria per species
LGT.bac2 <- LGT.bacteria %>% count(.id, sort = T)
LGT.bac <- na.omit(LGT.bac)
LGT.bac2 <- na.omit(LGT.bac2)

# Merge both datasets to plot the bacteria as relative count
LGT.bac<-merge(LGT.bac,LGT.bac2,by=".id",all.x=T)
colnames(LGT.bac)<-c("GAGAid","prokaryote_origin","count","totalCount")

# Calculate the relative bacterial count of e.g Wolbachia (for example 80 % Wolbachia, 20 % Blochmannia)
LGT.bac$relative<-LGT.bac$count/LGT.bac$totalCount
```

# A.6) Add the bacterial origin
```{r}
tp2 <- tp1 +
  new_scale_fill() +
  geom_fruit(
          data=LGT.bac,
          geom=geom_bar,
          mapping=aes(y=GAGAid, x=relative,fill=prokaryote_origin), #The count will be mapped to x
          pwidth=.1,      # Width of the external layer
          stat="identity",
          orientation="y", #The orientation of the axis
          position=position_stackx(hexpand=hex2),   #Adjust the position of the geom_layer
          axis.params =list(
                          axis="x",       # add axis text of the layer.
                          vjust=0.2,
                          text.angle=0,   # the text size of axis.
                          text.size=1.5,
                          hjust=1,        # adjust the horizontal position of text of axis
                          nbreak=1,       # number of lines within geom_layer,
                          ),
          grid.params=list(alpha=.2)      # Parameter to adjust gridlines around the tree
      ) +
      scale_fill_brewer(    #color of the external geom_layer
      
       name="HGT Prokaryote origin",        #name of the legend for the external geom_layer
       guide=guide_legend(keywidth=1,keyheight=1,order=6, ncol=6)
      ) +
      theme(legend.position="bottom", # the position of legend.
            legend.box="vertical", legend.margin = margin(0,0,0,0), #change the spacing between the legend and the plot
            legend.box.margin=margin(20,20,20,20), #change the spacing between the legend and the plot
          legend.background=element_rect(fill=NA), # the background of legend.
          legend.title=element_text(size=12), # the title size of legend.
          legend.text=element_text(size=10) # the text size of legend.
         # legend.spacing.y = unit(0.01, "cm")
      )
tp2<-tp2+scale_fill_manual(values=c("grey20","lightgreen","grey80","grey90","grey60"))

tp2

ggsave("./plots/GAGA.new.phylo.HGTs.pdf", width=20, height=25)
```


```{r}
# Join dataframes for LGT representation
LGT.complete<-LGTs %>% 
  full_join(y=LGT.bacteria, by=c("locusID",".id","species","subfamily"))

# Change order
LGT.complete<-LGT.complete %>% 
  relocate(species, subfamily, Protein.names,prokaryote_origin,order_prokaryote_origin,class_prokaryote_origin, .after = .id)

# Save dataframe as supplementary table
write.csv(LGT.complete, "./0_data/Tables/497_HQ_LGTs_all_info.csv", row.names=FALSE)
write.xlsx(LGT.complete,"./0_data/Tables/497_HQ_LGTs_all_info.xlsx")
```


# B.1) Pre-curated LGT data: LGT candidates after automated filtering and before manual curation & prokaryotic gene annotation
After automated filtering of all 13,664 predicted LGT candidates, 1,148 HGT candidates across XY genomes remained for further manual curation and prokaryotic gene annotation, including also the stLFR genomes.
```{r}
# Load data
preCurated.LGTs<-read_excel("./0_data/Tables/GAGA.HGTs.afterFiltering.all.genomes.xlsx")

# Fit data format to 497 HQ-HGT table
## add GAGA-id in locusID column
preCurated.LGTs$locusID<-paste(preCurated.LGTs$.id, preCurated.LGTs$locusID, sep = ".")

## bring column locusID into the following format: GAGA-id.Scaffold.start-end
preCurated.LGTs$locusID <- gsub(":", ".", preCurated.LGTs$locusID)

## Count number of HGT CDS sequences per HGT locus
preCurated.LGTs$nr_predicted_loci <- str_count(preCurated.LGTs$cand.start, ";") + 1

## For Analysis: Count total number of HGT candidates and species across 1,148 pre-curated HGT loci
HGT_CDS_sum <- sum(preCurated.LGTs$nr_predicted_loci)
preCurated.LGTs$species<-as.factor(preCurated.LGTs$species)
HGT_species <- levels(preCurated.LGTs$species)

write.xlsx(preCurated.LGTs,"./Supplementary_Tables/GAGA.HGTs.afterFiltering.all.genomes.xlsx")
```

# B.2) Pre-curated LGT data: Add the LGT candidate numbers per genome to the phylogeny
Plot the circular phylogeny and add red dot to the tip, where the size of the red dot represents the number of identified pre-curated LGTs.
```{r}
#Create a tally dataframe that counts how many LGT candidates we have per species.
preCurated.LGT.tally<-as.data.frame(table(select(preCurated.LGTs,.id)))
colnames(preCurated.LGT.tally)<-c("GAGAid","LGT.count")

#Add the LGT.tally dataframe to the phylogeny object as additional data.

pl1<-p0 %<+% df +
  geom_tiplab(aes(label=Species),
  offset=10,
  size=3) +
  theme(legend.position="bottom")
pl1

# Plot node labels to color subfamilies accordingly
pl1 + geom_text(aes(label=node))


# Define nodes to color subfamilies based on
nodes <- data.frame(node=c(250,1, 167, 188,192, 30, 195, 200, 206, 207, 210, 249), type = c("Myrmicinae", "Leptanillinae", "Ponerinae", "Proceratiinae", "Amblyoponinae", "Paraponerinae", "Dorylinae", "Dolichoderinae", "Pseudomyrmecinae", "Myrmeciinae", "Formicinae", "Ectatomminae"))

# Highlight subfamilies by color
pl2 <- pl1 +
  geom_hilight(data=nodes, aes(node=node, fill=type), type = "gradient", alpha = .3) +
  scale_color_gradient()
pl2


plot1<-pl2 %<+% preCurated.LGT.tally

#Plot the phylogeny and add red dot to the tip, where the size of the red dot represents the number of identified LGTs
plot2<-plot1+geom_tippoint(aes(size=LGT.count),col="red", alpha=.5)
plot2

ggsave("./Supplementary_Figures/GAGA.preCurated.HGTs.pdf")
````

