---
title: "ANK repeat proteins"
author: "Janina Rinke"
date: "`r format(Sys.time(), '%d %B, %Y')`"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r libraries, include = FALSE}
#if (!requireNamespace("BiocManager", quietly = TRUE))
   #install.packages("BiocManager")

#BiocManager::install("treeio")
#install.packages("devtools")
#devtools::install_github("GuangchuangYu/ggtree")
library(treeio)
library(ggtree)
library(ggplot2)
library(phytools)
library(viridisLite)
library(googlesheets4)
library(dplyr)
library(knitr)
library(tidyverse)
library(kableExtra)
library(flextable)
library(magrittr)

#BiocManager::install("ggstar")
#BiocManager::install("ggtreeExtra")
library(ggtreeExtra)
library(ggnewscale)
library(tidytree)
library(ggstar)
library(viridis)
library(ggstance)
library(reshape2)
```

### Load the LGT table with all high-quality candidates
```{r}
library(readxl)

# Read CDS level table to extract all unique HGTs
LGTs<-read_excel("/Users/Janina/sciebo/MASTER/Masterarbeit/lgt/0_data/LGTs.CDSLevel.UniProt.Info.Reannotated.xlsx")
```



### Subset the dataframe to only include ankyrin repeat-domain containing proteins
```{r}
colnames(LGTs)

# Only keep ankyrin-repeat proteins
ANKs<-LGTs %>%
  dplyr::filter(grepl('Ank|OTU|Tetratricopeptide repeat protein', Protein.names) | grepl('ANKYR', CDS.cogs))

# Rename the uncharacterized proteins
ANKs$Protein.names<-gsub("Uncharacterized protein","Uncharacterized ANK",ANKs$Protein.names)

# Count number of different ankyrin proteins per species
Nr.ANKs.per.species<-ANKs %>% 
  dplyr::count(.id,species, subfamily, CDS.UniprotIDs,Protein.names)

# Rename column
Nr.ANKs.per.species<- Nr.ANKs.per.species%>%
  dplyr::rename("count" = "n")
```



### Ankyrin bubble plot
```{r}
# Most basic bubble plot
ggplot(ANKs)+geom_point(aes(size=as.numeric(CDS.lengths),
                            y=as.numeric(Reannotated.CDS.pids),x=as.numeric(Reannotated.s_covs),
                            fill=Protein.names,col=subfamily),pch=21,stroke=1)
```



### Plot the ankyrin repeats based on the UniProt IDs onto the GAGA ant phylogeny

Load data from google spreadsheet (Already stored in an R object that you can just load)
```{r}
#id<-"1e7Hj_BQ8rke5XW7j93YL6QVAGbcvsb2yw7oD86WnPJY"
#gsheet<-read_sheet(id,sheet="GlobalOverview",skip = 2,na = c(""," ","NA","#DIV/0!"),col_types = "c",trim_ws=T)
#GlobalOverview<-as.data.frame(gsheet,stringsAsFactors=F)
#save(GlobalOverview,file="../../misc/data/gsheet.GlobalOverview.21-09-22.Robj")
load("/Users/Janina/sciebo/MASTER/Masterarbeit/lgt/0_data/gsheet.GlobalOverview.21-09-22.Robj")
```

Create a dataframe where the first row is the GAGA id (e.g. GAGA-0002), which are also used as the tip labels in the phylogeny
```{r}
data2add<-data.frame(GAGA.id = GlobalOverview$`GAGA ID`, GlobalOverview)
```

Gather subfamily info for NCBI IDs as well and manually add the subfamily
```{r}
colnames(data2add)

subset(data2add,is.na(Subfamily))

data2add$Subfamily[data2add$GAGA.id=="NCBI-0001"]<-"Dorylinae"
data2add$Subfamily[data2add$GAGA.id=="NCBI-0002"]<-"Myrmicinae"
data2add$Subfamily[data2add$GAGA.id=="NCBI-0003"]<-"Myrmicinae"
data2add$Subfamily[data2add$GAGA.id=="NCBI-0004"]<-"Myrmicinae"
data2add$Subfamily[data2add$GAGA.id=="NCBI-0005"]<-"Formicinae"
data2add$Subfamily[data2add$GAGA.id=="NCBI-0006"]<-"Myrmicinae"
data2add$Subfamily[data2add$GAGA.id=="NCBI-0007"]<-"Ponerinae"
data2add$Subfamily[data2add$GAGA.id=="NCBI-0008"]<-"Formicinae"
data2add$Subfamily[data2add$GAGA.id=="NCBI-0009"]<-"Ponerinae"
data2add$Subfamily[data2add$GAGA.id=="NCBI-0010"]<-"Dolichoderinae"
data2add$Subfamily[data2add$GAGA.id=="NCBI-0011"]<-"Ponerinae"
data2add$Subfamily[data2add$GAGA.id=="NCBI-0012"]<-"Myrmicinae"
data2add$Subfamily[data2add$GAGA.id=="NCBI-0013"]<-"Myrmicinae"
data2add$Subfamily[data2add$GAGA.id=="NCBI-0014"]<-"Myrmicinae"
data2add$Subfamily[data2add$GAGA.id=="NCBI-0015"]<-"Myrmicinae"
data2add$Subfamily[data2add$GAGA.id=="NCBI-0016"]<-"Myrmicinae"
data2add$Subfamily[data2add$GAGA.id=="NCBI-0017"]<-"Myrmicinae"
data2add$Subfamily[data2add$GAGA.id=="OUT-0001"]<-"Formicinae"
data2add$Subfamily[data2add$GAGA.id=="OUT-0002"]<-"Formicinae"
```

Create a column with clean species IDs for all species.

This has to be done, because the database of all the sequenced species is a bit convoluted.
The function coalesce() finds the first non-missing value at each position. In this case, the columns "Species" and "Species..from.Assembly.Overview" are collapsed. If the column "Species" does not have an entry, the entry from the column "Species..from.Assembly.Overview" will be taken.
```{r}
data2add$cleanSpeciesID<-dplyr::coalesce(data2add$Species,data2add$Species..from.Assembly.Overview.)
```


Load the tree file of all the GAGA genomes.
```{r}
tree <- read.iqtree("/Users/Janina/sciebo/MASTER/Masterarbeit/lgt/0_data/hymenoptera_psc98_partition_b1000_40threads_msubnucl.treefile")
```

Create a very quick and dirty plot of the phylogeny
```{r}
plot(tree@phylo,show.node.label = T,show.tip.label = T,cex=.5)
nodelabels(bg=rgb(0,0,0,0),col="red",frame = "none")
```

Check the plot to see where the root should be. Here we root the tree at node 166 that separates the outgroups, the honeybee "Amel" (Apis mellifera) and the parasitoid wasp Nasonia (Nvit), from ants.
```{r}
#tree.rooted<-treeio::root(tree,node=166)
tree.rooted <- tree
tree.rooted.subset<-treeio::drop.tip(tree.rooted,tip=c("NCBI-0018_Amel","NCBI-0019_Nvit"))
```

Plot basic tree
```{r}
tp<-ggtree(tree.rooted.subset, branch.length = "none") #+geom_rootedge(.01)
```

Add the species identity from `FAS.genes.and.names.tsv`
```{r}
tp$data$label[tp$data$isTip==T]<-gsub("_.*","",tp$data$label[tp$data$isTip==T])

tp2<-tp %<+% data2add
tp2 
```

```{r}
#install.packages("ggnewscale")
library("ggnewscale")

#Make Subfamily as a factor (discrete color scale for factors)
tp2$data$Subfamily<-as.factor(tp2$data$Subfamily)

# tp2.5 is a less complicated tree without UFboot
tp2.5<-tp2 +
  geom_tiplab(aes(label=paste(cleanSpeciesID,sep=" "), color = Subfamily),size=2) # remove GAGAid from label
```

# Plot the tree
```{r}
tp2.5
```

### Plot ankyrins onto phylogeny
```{r}
library(pals)
library(RColorBrewer)

tp3 <- tp2.5 %<+% Nr.ANKs.per.species
tp3

p2<-facet_plot(tp3,data=Nr.ANKs.per.species, mapping=aes(x=Nr.ANKs.per.species$count, fill=Nr.ANKs.per.species$Protein.names), geom = geom_barh, stat="identity", panel="Nr of ANKs") + hexpand(.2, direction = 1) +
    theme(legend.position = 'bottom',
    legend.key = element_blank(), # removes the border
    legend.key.size = unit(2, 'cm'), # sets overall area/size of the legend
    legend.text = element_text(size = 20)) +
  theme_bw()

colourCount = length(unique(Nr.ANKs.per.species$Protein.names))
getPalette = colorRampPalette(brewer.pal(17, "PuOr"))
p2
 

p3<-p2 +
  guides(shape = guide_legend(order = 1),col = guide_legend(order = 2), fill=guide_legend(title="Ankyrin type")) +
     scale_fill_manual(values = getPalette(colourCount)) +
  theme(legend.key.size=unit(0.5, "cm"),
        legend.text = element_text(size=10))
    #scale_fill_viridis(discrete = TRUE)

p3

# Export plot w ggsave
ggsave(
  "/Users/Janina/sciebo/HGT_Manuscript/Figures/Anks_Phylo.pdf",
  plot = p3, 
  device = "pdf", 
  dpi=150, 
  width=65, 
  height=75, 
  units = "cm"
  )
```
## Prepare a tsv file for the ANK metagenomic analysis 
```{r}
## For CDS sequences 
# Only keep selected columns
ANKs.df<- ANKs %>%
  select(.id, locus.scaffold, Reannotated.CDS.starts, Reannotated.CDS.ends, CDS.UniprotIDs, Protein.names, Organism)

# Remove column headers
names(ANKs.df)<-NULL

write.table(ANKs.df, file='/Users/Janina/sciebo/HGT_Manuscript/Dataframes/GAGA.ANKs.tsv', quote=FALSE, sep='\t', row.names = F)

## For ANK loci
# Only keep selected columns
ANKs.loci.df<- ANKs %>%
  select(.id, locus.scaffold, cand.start, cand.end)

# Only keep distinct rows
ANKs.loci.df <- ANKs.loci.df %>%
  distinct()

# Remove column headers
names(ANKs.loci.df)<-NULL

write.table(ANKs.loci.df, file='/Users/Janina/sciebo/HGT_Manuscript/Dataframes/GAGA.ANKs.loci.txt', quote=FALSE, sep='\t', row.names = F)
```

## Separate dataframes into different categories
1: LGTs CDS level (1053 CDS)
2: HGTs Locus level (497 loci)
3: Ankyrins (417 CDS)
4: Species-specific LGTs CDS (101 Loci)
5: Species-specific newly-filtered (88 loci)
```{r}
# Rename columns which are wrongly named
write.table(ANKs, file='/Users/Janina/sciebo/HGT_Manuscript/Dataframes/ANKs.tsv', quote=FALSE, sep=',', row.names = F)
write.table(Unique.LGTs, file='/Users/Janina/sciebo/HGT_Manuscript/Dataframes/Species-specific_LGTs.newly.filtered.tsv', quote=FALSE, sep=',', row.names = F)
write.table(Expressed.Unique.LGTs, file='/Users/Janina/sciebo/HGT_Manuscript/Dataframes/Expressed.Unique.LGTs.tsv', quote=FALSE, sep=',', row.names = F)
```

