---
title: "Other HGTs"
author: "Janina Rinke"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
    code_folding: hide
  pdf_document: default
--- 
</STYLE>
---

```{r global-options, include=FALSE}
# Define the global chunk options
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, message=FALSE)
```

```{r libraries, include = FALSE}
library(knitr)
library(tidyverse)
library(kableExtra)
library(flextable)
library(magrittr)
library(dplyr)
library(ggplot2)
library(mishMashr)
```

## Load the LGT table with all high-quality CDS candidates
```{r}
library(readxl)

# Read CDS level table to extract all unique LGTs later
LGTs<-read_excel("./0_data/Tables/LGTs.CDSLevel.UniProt.Info.Reannotated.xlsx")
```



## Group all LGTs based on protein type. 
Exclude all convergently transferred ankyrin-related proteins, as well as ancient, orthologous LGTs which have been analysed elsewhere (Lysozymes, MurNAc etherases & CFA Synthases). Additionally, all uncharacterized & fragmented proteins, and TE-related LGTs are excluded from further in-depth analysis. 
```{r}
colnames(LGTs)

# create grouping for protein type
LGTs$type<-"other"
LGTs$type[grep("Ankyrin|OTU|ankyrin|Tetratricopeptide repeat protein|TPR repeat-containing protein|Reverse transcriptase domain-containing protein|Phosphocholine_transferase_AnkX",LGTs$Protein.names)]<-"Ank"
LGTs$type[grep("Cyclopropane|EC 2.1.1.79",LGTs$Protein.names)]<-"CFA"
LGTs$type[grep("N-acetylmuramic",LGTs$Protein.names)]<-"MurNAc"
LGTs$type[grep("DUF1016|Lysozyme|lysozyme|Lyzozyme",LGTs$Protein.names)]<-"Lys"
LGTs$type[grep("Uncharacterized|uncharacterized|unknown|Deleted",LGTs$Protein.names)]<-"Unknown"
LGTs$type[grep("Fragment",LGTs$Protein.names)]<-"Fragmented"
LGTs$type[grep("Transposase|transposase|DNA repair protein|Helix-turn-helix transcriptional regulator",LGTs$Protein.names)]<-"TE-related"
LGTs$type[grep("50S",LGTs$Protein.names)]<-"Ribosomal protein"

# Remove all clade-specific HGTs like ankyrins, lysozymes, MurNAcs, unchacracterized LGTs..
other.LGTs<-LGTs %>%
  dplyr::filter(type=="other") %>%     # Only keep LGTs of the catecory "other"
  dplyr::filter(!is.na(Protein.names)) # Remove all LGTs without a Uniprot annotation

CFA.LGTs<-LGTs %>%
  dplyr::filter(type=="CFA")     # Only keep LGTs of the category "other"

library("xlsx")
write.xlsx(CFA.LGTs,"./Data/CFA_Synthases/CFA_HQ_HGTs.xlsx")
```
Removing all LGTs without a Uniprot annotation, all Ankyrins, Lysozymes, MurNAc etherases, CFA Synthases and TE-related LGTs, as well as uncharacterized and fragmented LGTs, results in 125 remaining LGTs of the category "other".



## General investigation of "other" LGTs
```{r}
# Count the proteins which occur only once in the dataframe both based on the column "Protein.names" and   "CDS.UniprotIDs" to be entirely sure to remove all duplicated entries
Unique.LGTs<-once_only(other.LGTs, "Protein.names", return_column_only = F)
Unique.LGTs<-once_only(Unique.LGTs, "CDS.UniprotIDs", return_column_only = F)
# 61 LGT CDS occurring only once


# Count number of HGT loci with unique LGTs
Unique.loci.count<-Unique.LGTs %>% 
  dplyr::count(locusID)
# 45 HGT loci with unique LGTs


# Count only proteins which occur more than once in the dataframe in at least two species
multiple.LGTs<-other.LGTs %>%
  group_by(Protein.names) %>%
  filter(n_distinct(species) >= 2)
# 32 occasions of multiple LGTs

# Count number of species with other proteins
Other.species.count<-other.LGTs %>% 
  dplyr::count(species, subfamily)
# 43 species from seven subfamilies with other proteins


# Count the number of bacterial donors
other.LGTs$donor<-"other"
other.LGTs$donor[grep("Sodalis",other.LGTs$Organism)]<-"Sodalis"
other.LGTs$donor[grep("Wolbachia",other.LGTs$Organism)]<-"Wolbachia"
other.LGTs$donor[grep("Serratia",other.LGTs$Organism)]<-"Serratia"
other.LGTs$donor[grep("Spiroplasma",other.LGTs$Organism)]<-"Spiroplasma"
other.LGTs$donor[grep("Yersinia",other.LGTs$Organism)]<-"Yersinia"
other.LGTs$donor[grep("Cardinium",other.LGTs$Organism)]<-"Cardinium"
other.LGTs$donor[grep("Baumannia|Blochmannia|Candidatus",other.LGTs$Organism)]<-"Blochmannia-like"

Other.bac.count<-other.LGTs %>% 
  dplyr::count(donor)
colnames(Other.bac.count)<-c("donor","count")

head(other.LGTs)
```
Considering only uniquely occurring HGTs results in 61 unique CDS in 45 loci. 43 ant species from seven subfamilies contained at least one uniquely occurring LGT. Bacterial donors were most often Wolbachia (n=89), followed by Spiroplasma (n = 13), Sodalis (n=9), Cardinium (n=9), Blochmannia-like (n=6) and other bacteria.



## In-depth analysis of other LGTs: Only keep expressed proteins (read count > 100)
```{r}
# Count the number of expressed "other" LGTs
other.LGTs$CDS.totalcounts<-as.numeric(other.LGTs$CDS.totalcounts)
other.LGTs$Reannotated.CDS.totalcounts<-as.numeric(other.LGTs$Reannotated.CDS.totalcounts)

# RNAseq counts first annotation
other.LGTs %>% 
  dplyr::count(CDS.totalcounts > 100)
#---------------------------------#
#CDS.totalcounts > 100
#FALSE	105			
#TRUE	18			
#NA	2	
#----------------------------------#


# RNAseq counts re-annotation of candidates
other.LGTs %>% 
  dplyr::count(Reannotated.CDS.totalcounts > 100)
#---------------------------------#
#Reannotated.CDS.totalcounts > 50
#FALSE 95				
#TRUE	17			
#NA	13
#----------------------------------#

# In depth analysis of expressed and full-length "other" LGTs
other.LGTs$CDS.s_covs<-as.numeric(other.LGTs$CDS.s_covs)
other.LGTs$Reannotated.s_covs<-as.numeric(other.LGTs$Reannotated.s_covs)
other.LGTs$Reannotated.CDS.pids<-as.numeric(other.LGTs$Reannotated.CDS.pids)

# Filter all of these "other" LGTs for expressed & (almost) full-length genes
Expressed.full.other.LGTs<-other.LGTs %>%
  dplyr::filter(CDS.totalcounts > 100 & Reannotated.CDS.totalcounts > 100) %>% 
  dplyr::filter(Reannotated.s_covs >= 80) %>%
  dplyr::filter(Reannotated.CDS.pids > 65.0)

# Remove duplicated entry for GAGA-0080, as this is a wrong annotation
Expressed.full.other.LGTs<-Expressed.full.other.LGTs %>%
  dplyr::filter(CDS.UniprotIDs != "A0A218KN31")
```
Keeping only expressed and close to full-length (subject coverage > 80 %) "other" LGTs resulted in 6 remaining CDS for in-depth analysis.



```{r}
# Assign general functions
Unique.LGTs$funct<-"other"
Unique.LGTs$funct[grep("transcriptional|Transcription|transcription",Unique.LGTs$Protein.names)]<-"Transcription"
Unique.LGTs$funct[grep("transposase|Transposase",Unique.LGTs$Protein.names)]<-"TE-related"
Unique.LGTs$funct[grep("transport|Transport",Unique.LGTs$Protein.names)]<-"Transport"
Unique.LGTs$funct[grep("biosynthesis|de novo",Unique.LGTs$Protein.names)]<-"Metabolic function"
Unique.LGTs$funct[grep("biosynthetic|metabolic",Unique.LGTs$Gene.ontology..biological.process.)]<-"Metabolic function"
Unique.LGTs$funct[grep("regulation",Unique.LGTs$Gene.ontology..biological.process.)]<-"Regulatory function"
Unique.LGTs$funct[grep("protein",Unique.LGTs$Gene.ontology..biological.process.)]<-"Protein-related function"
Unique.LGTs$funct[grep("cell wall | membrane",Unique.LGTs$Gene.ontology..biological.process.)]<-"Cell wall/Membrane-related"
Unique.LGTs$funct[grep("cell wall | membrane",Unique.LGTs$Gene.ontology..GO.)]<-"Cell wall/Membrane-related"
Unique.LGTs$funct[grep("DNA|RNA",Unique.LGTs$Gene.ontology..molecular.function.)]<-"DNA repair / DNA binding"
Unique.LGTs$funct[grep("metal",Unique.LGTs$Gene.ontology..molecular.function.)]<-"Metallopeptidase activity"

Funct<-Unique.LGTs %>% 
  dplyr::count(funct)
colnames(Funct)<-c("Function","count")

Unique.LGTs$Reannotated.s_covs<-as.numeric(Unique.LGTs$Reannotated.s_covs)
```



```{r}
##### Eight species-specific LGTs remain for in-depth analysis.###########

# Plot GO terms for each expressed unique LGT
library(UniprotR)

####### p1
prot_info<-GetProteinGOInfo("A0A060Q018")
p1<-PlotGOAll(GOObj = prot_info, Top = 10, directorypath = getwd(), width = 8, height = 5)
# Customize the title with the protein ID
p1 <- p1 + 
  ggtitle(paste("UDP-N-acetylglucosamine 1-carboxyvinyltransferase"))
p1

####### p2
prot_info<-GetProteinGOInfo("W0HVG4")
p2<-PlotGOAll(GOObj = prot_info, Top = 10, directorypath = getwd(), width = 8, height = 5)
# Customize the title with the protein ID
p2 <- p2 + 
  ggtitle(paste("D-alanine--D-alanine ligase"))
p2

####### p3
prot_info<-GetProteinGOInfo("A0A1A9NKN6")
p3<-PlotGOAll(GOObj = prot_info, Top = 10, directorypath = getwd(), width = 8, height = 5)
# Customize the title with the protein ID
p3 <- p3 + 
  ggtitle(paste("Phosphoribosylformylglycinamidine cyclo-ligase "))
p3

####### p4
prot_info<-GetProteinGOInfo("A0A060Q4G1")
p4<-PlotGOAll(GOObj = prot_info, Top = 10, directorypath = getwd(), width = 8, height = 5)
# Customize the title with the protein ID
p4 <- p4 + 
  ggtitle(paste("DNA mismatch repair protein "))
p4

####### p4
prot_info<-GetProteinGOInfo("A0A7C8GK68")
p4<-PlotGOAll(GOObj = prot_info, Top = 10, directorypath = getwd(), width = 8, height = 5)
# Customize the title with the protein ID
p4 <- p4 + 
  ggtitle(paste("DNA mismatch repair protein "))
p4
```




### Make a multi-panel plot
```{r}
library(ggpubr) #load in library for multi-panel figures
library(gridExtra)

p1 <- p1 + theme(plot.margin = unit(c(1, 1, 1, 1), "cm"))
p2 <- p2 + theme(plot.margin = unit(c(1, 1, 1, 1), "cm"))
p3 <- p3 + theme(plot.margin = unit(c(1, 1, 1, 1), "cm"))
p4 <- p4 + theme(plot.margin = unit(c(1, 1, 1, 1), "cm"))

#put all three plots together into one multipanel plot
multi_plot<- ggarrange(p1,p2,p3,p4, #plots that are going to be included in this multipanel figure
                       labels = c("A", "B", "C","D"), #labels given each panel 
                       ncol = 2, nrow = 2, #adjust plot space 
                       common.legend = T #does the plot have a common legend
                       )
#add titles and labels to the multi-panel graph
multi_plot <- annotate_figure(multi_plot,
                              top = text_grob("GO terms", color = "black", face = "bold", size = 11))
multi_plot 

ggsave("./plots/Unique.LGTs.4plots.pdf", multi_plot, width=15, height=10)
```




### Make different plots to provide an overview of species-specific HGTs
```{r}
install.packages("ggpubr")
library(ggpubr)

# Make a plot to show how many species have unique LGTs and they subfamily identity
species_bp <- ggbarplot(Unique.species.count, x = "species", y = "n",
          fill = "subfamily",               # change fill color by cyl
          color = "white",            # Set bar border colors to white
          palette = "jco",            # jco journal color palett. see ?ggpar
          sort.val = "asc",           # Sort the value in ascending order
          sort.by.groups = TRUE,      # Sort inside each group
          x.text.angle = 60           # Rotate vertically x axis texts
          ) +
  ylab("Number of unique HGTs") +
  xlab("")
species_bp + font("x.text", size = 10)

#ggsave("/Users/Janina/Documents/MASTER/Masterarbeit/plots/Thesis_Plots/Unique.LGTs.of.species_RAW.pdf", species_bp, width=15, height=10)

# Make a plot to show the expression 
Unique.LGTs$Reannotated.CDS.lengths<-as.numeric(Unique.LGTs$Reannotated.CDS.lengths)
Unique.LGTs$Reannotated.s_covs<-as.numeric(Unique.LGTs$Reannotated.s_covs)

Unique.LGTs$Uniprot_organism<-gsub("cellular.*?),","",Unique.LGTs$Uniprot_organism)
Unique.LGTs$Uniprot_organism<-gsub(".*?, .*?, .*?, (.*?), .*","\\1",Unique.LGTs$Uniprot_organism)

expression_vs_s_cov <- ggscatter(Unique.LGTs, x = "Reannotated.s_covs", y = "Reannotated.CDS.totalcounts",
                add = "reg.line",               # Add regression line
                conf.int = TRUE              # Add confidence interval
                #color = "Uniprot_organism", palette = "jco", # Color by groups "cyl"
              # shape = "Uniprot_organism"                   # Change point shape by groups "cyl"
                )
expression_vs_s_cov

head(Unique.LGTs)

Unique.LGTs$Reannotated.CDS.totalcounts<-as.numeric(Unique.LGTs$Reannotated.CDS.totalcounts)
# Make a density plot to show s_cov
s_cov<-Unique.LGTs %>%
 #filter( Reannotated.CDS.totalcounts<5000 ) %>%
  ggplot( aes(x=Reannotated.s_covs)) +
    geom_density(fill="#69b3a2", color="#e9ecef", alpha=0.8) +
    theme_minimal()
s_cov
```


## Make a plot for the bacterial donors
```{r}
# Compute percentages
Unique.bac.count$fraction <- Unique.bac.count$count / sum(Unique.bac.count$count)

# Compute the cumulative percentages (top of each rectangle)
Unique.bac.count$ymax <- cumsum(Unique.bac.count$fraction)

# Compute the bottom of each rectangle
Unique.bac.count$ymin <- c(0, head(Unique.bac.count$ymax, n=-1))

# Compute label position
Unique.bac.count$labelPosition <- (Unique.bac.count$ymax + Unique.bac.count$ymin) / 2

# Compute a good label
Unique.bac.count$label <- paste0(Unique.bac.count$donor, "\n n= ", Unique.bac.count$count)

# Make the plot
bac.plot<-ggplot(Unique.bac.count, aes(ymax=ymax, ymin=ymin, xmax=4, xmin=3, fill=donor)) +
  geom_rect() +
  geom_label( x=3.5, aes(y=labelPosition, label=label), size=6) +
  scale_fill_brewer(palette=4) +
  coord_polar(theta="y") +
  xlim(c(2, 4)) +
  theme_void() +
  theme(legend.position = "none")
bac.plot

#ggsave("./plots/Unique.LGTs.bacteria_RAW.pdf", bac.plot, width=10, height=10)
```

# Plot GO terms for Unique LGTs only
```{r}
library(UniprotR)
go<-GetProteinGOInfo(Unique.LGTs$CDS.UniprotIDs[!is.na(Unique.LGTs$CDS.UniprotIDs)])
#Plot Biological process information top 10 go terms  
PlotGOBiological(go, Top = 10) 
#Plot molecular function information top 20 go terms
Plot.GOMolecular(go, Top = 20)
#Plot subcellualr localization information 
Plot.GOSubCellular(go) 
#Combine Gene ontology plots into one plot 
p_combined<-PlotGoInfo(go)
#Handy visualization for publications 
p2_combined<-PlotGOAll(GOObj = go, Top = 10, directorypath = getwd(), width = 8, height = 5)
```


## Make a plot concerning function and identity
```{r}
head(Funct)

Funct$Function<-as.factor(Funct$Function)

Funct %>%
  mutate(Function = fct_reorder(Function, desc(count))) %>%
  ggplot( aes(x=Function, y=count)) +
    geom_bar(stat="identity", fill="#69b3a2", alpha=.6, width=.4) +
    coord_flip() +
    xlab("") +
    theme_minimal()

#ggsave("./plots/Unique.LGTs.function_RAW.pdf", width=12, height=10)
```
