---
title: "Calculate expression of single CDS of LGT regions"
author: "Janina Rinke"
output: 
  html_document:
    toc: true
    toc_float: true
    toc_depth: 2
    collapsed: true
    number_sections: true
    df_print: paged
  pdf_document: default
---
<STYLE TYPE="text/css">
<!--
  td{
    font-family: Arial;
    font-size: 8pt;
    padding:0px;
    cellpadding="0";
    cellspacing="0"
  }
  th {
    font-family: Arial;
    font-size: 8pt;
    height: 20px;
    font-weight: bold;
    text-align: right;
    background-color: #ccccff;
  }
  table {
    border-spacing: 0px;
    border-collapse: collapse;
  }
--->
</STYLE>
---

This script combines the information about all LGT candidates and respective CDS detected by dfast together with the RNAseq mapping expression. The read counts for every single CDS are added to the dataframe and plotted for the candidates.

```{r global-options, include=FALSE}
# Define the global chunk options
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, message=FALSE)
```

```{r libraries}
library(ggplot2)
library(data.table)
library(dplyr)
library(tidyr)
```

## Extract expression from RNAseq mapping files

#### Load the RNAseq mapping files for every LGT candidate into R.
```{r load files}
#Mac Computer:
#txt2load<-"/Users/Janina/sciebo/Master.LGTs/RNAseqmapping/*/RNAseq_mapping_count_summary.txt"
#Linux Computer
txt2load<-"/home/j/j_rink02/sciebo/Master.LGTs/RNAseqmapping/*/RNAseq_mapping_count_summary.txt"
files.to.load<-Sys.glob(txt2load)
dataFiles <- lapply(files.to.load,read.delim2,sep = "\t")
names(dataFiles)<-gsub("/home/j/j_rink02/sciebo/Master.LGTs/RNAseqmapping/(.*?)\\.fa.*","\\1",files.to.load)
```
Here, `r length(dataFiles)` RNAseq mapping files were loaded. Some candidates did not have RNAseq data, therefore some files are missing.

To check, whether all dataframes have been imported correctly, you can view any element from the list:
```{r check files}
summary(dataFiles[1])
dataFiles[1]
```


```{r load single file, include=FALSE}
#dataFiles[["GAGA-0515.Scaffold8.1760778-1761241"]]
i<-20
#txtfile <- read.delim2("/home/j/j_rink02/sciebo/Master.LGTs/RNAseqmapping/GAGA-0515.Scaffold8.1760778-1761241.fa/RNAseq_mapping_count_summary.txt", header = TRUE, sep = "\t", dec = ",")
txtfile<-dataFiles[[i]]
```


#### Calculate the sum of reads per CDS (sum of all samples e.g. queens, workers, males)
```{r sum-expression}
df<-txtfile %>%                               # Specify data frame
  dplyr::group_by(chr,start,end) %>%          # Specify group indicator
  dplyr::summarise_at(vars(ucount),           # Specify column
               list(totalcount = sum))        # Specify function

df$locusID<-names(dataFiles)[i]
df<-df %>% 
  separate(locusID, sep="\\.", c(".id","locus.scaffold","locus.position")) %>% 
  dplyr::rename(CDS.start=start, CDS.end=end)

```

```{r}
df.dfast<-read.csv("/home/j/j_rink02/sciebo/Master.LGTs/dfast/GAGA-0025.Scaffold15.4180270-4181120.fa/genome.gff_mod.gff",sep="\t",header=F)
df.new<-merge(df, df.dfast,by.x=c("chr","CDS.start","CDS.end"),by.y=c("V1","V4","V5"),all.x=T,all.y=T)

df.new$product<-gsub(".*product=(.*?)\\;.*","\\1",df.new$V9)

# relevant only when sumarizing a CDS-level table to LGT-level
df.final<-df.new %>%  
  dplyr::group_by(.id, locus.scaffold, locus.position) %>% 
  dplyr::summarize(CDS.starts = paste0(CDS.start, collapse = ";"),CDS.ends = paste0(CDS.end, collapse = ";"),CDS.totalcounts = paste0(totalcount, collapse = ";"),CDS.products = paste0(product, collapse = ";")) 

# To extract pid, q_cov_s_cov
gsub(".*?\\[(.*?)\\].*","\\1",df.dfast$V9)
```


#### List the information in a table.
```{r make-table}

```

## Merge information for all LGTs together with expression CDS.
```{r load lgts-combined table}
library(readxl)
LGTs.combined<-read_excel("/Users/Janina/Documents/MASTER/Masterarbeit/lgt/0_data/LGTs.combined.dfast.xlsx")
```


`r nrow(LGTs.combined)` CDS were predicted by dfast.
