---
title: "Calculate expression of single CDS of LGT regions"
author: "Janina Rinke"
output: 
  html_document:
    toc: true
    toc_float: true
    toc_depth: 2
    collapsed: true
    number_sections: true
    df_print: paged
  pdf_document: default
---
<STYLE TYPE="text/css">
<!--
  td{
    font-family: Arial;
    font-size: 8pt;
    padding:0px;
    cellpadding="0";
    cellspacing="0"
  }
  th {
    font-family: Arial;
    font-size: 8pt;
    height: 20px;
    font-weight: bold;
    text-align: right;
    background-color: #ccccff;
  }
  table {
    border-spacing: 0px;
    border-collapse: collapse;
  }
--->
</STYLE>
---

This script combines the information about all LGT candidates and respective CDS detected by dfast together with the RNAseq mapping expression. The read counts for every single CDS are added to the dataframe and plotted for the candidates.

```{r global-options, include=FALSE}
# Define the global chunk options
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, message=FALSE)
```

```{r libraries}
library(ggplot2)
library(data.table)
library(dplyr)
library(tidyr)
library(purrr)
library(ape)
```

## Extract expression from RNAseq mapping files

#### Load the RNAseq mapping files for every LGT candidate into R.
```{r load files}
#Mac Computer:
#txt2load<-"/Users/Janina/sciebo/Master.LGTs/RNAseqmapping/*/RNAseq_mapping_count_summary.txt"
#Linux Computer
txt2load<-"/home/j/j_rink02/sciebo/Master.LGTs/RNAseqmapping/*/RNAseq_mapping_count_summary.txt"
files.to.load<-Sys.glob(txt2load) #Sys.glob() returns all paths to the files based on a wildcard pattern in our file paths
head(files.to.load)
```

Here, `r length(files.to.load)` RNAseq mapping files were loaded. Some candidates did not have RNAseq data, therefore some files are missing.

Save all file paths in a list called dataFiles and re-name the list elements
```{r load files 2}
dataFiles <- lapply(files.to.load,read.delim2,sep = "\t") #lapply function to read all files and save them as dataframes
#MacComputer
names(dataFiles)<-gsub("/Users/Janina/sciebo/Master.LGTs/RNAseqmapping/(.*?)\\.fa.*","\\1",files.to.load) # gsub() selectively replaces multiple occurrences of a text within an R string ("search term", "replacement term",string searched) 
#Linux Computer
names(dataFiles)<-gsub("/home/j/j_rink02/sciebo/Master.LGTs/RNAseqmapping/(.*?)\\.fa.*","\\1",files.to.load)
```

To check, whether all dataframes have been imported correctly, you can view any element from the list:
```{r check files}
summary(dataFiles[1])
dataFiles[1]
```


```{r load single file}
#dataFiles[["GAGA-0515.Scaffold8.1760778-1761241"]]
#i<-393
txtfile<-dataFiles[[393]]
head(txtfile)
```


#### Calculate the sum of reads per CDS (sum of all samples e.g. queens, workers, males) using dplyr
```{r sum-expression single file}
# For a single file
df<-txtfile %>%                               # Specify data frame
  dplyr::group_by(chr,start,end) %>%          # Specify indicators to group on 
  dplyr::summarise_at(vars(ucount),           # Specify column to summarize based on the group indicators
               list(totalcount = sum))        # Specify function (summarize as a list and save in column "totalcount")

df$locusID<-names(dataFiles)[393]             # Add the name of the LGT by adding a column called "locusID" 
df<-df %>% 
  separate(locusID, sep="\\.", c(".id","locus.scaffold","locus.position")) %>% 
  dplyr::rename(CDS.start=start, CDS.end=end)# Separate the locusID column into .id, scaffold and CDS position

df$chr<-NULL
head(df)
```

```{r sum-expression all}
#### For all files: 
#Define a function and apply function to all files with map()
summarizeExpression <- function(DF) {
  DF %>% 
 dplyr::group_by(chr,start,end) %>%          
 dplyr::summarise_at(vars(ucount),          
               list(totalcount = sum))
}

dataFiles2 <- map(dataFiles,summarizeExpression)

# Add LGT names to all dataframes as a column and separate into single columns with GAGAid, Scaffold and CDS coordinate.
#Check if all locusIDs have the structure {GAGA-id}.{Scaffold}.{start}-{stop}
names(dataFiles2)[grep("\\.1\\.",names(dataFiles2))] #Output: 22 Genomes with different structure e.g. "OUT-0001.WHNR01000088.1.32800-34183"


for (i in seq_along(dataFiles2)) {
  dataFiles2[[i]]$locusID <- gsub("\\.([0-9])\\.","_\\1.",names(dataFiles2)[i])
}

# separate the locusIDs in all files
separateLOCUSID <- function(DF) {
  DF %>% 
  separate(locusID, sep="\\.", c("GAGAid","locus.scaffold","locus.position")) %>% 
  dplyr::rename(CDS.start=start, CDS.end=end)  
}

dataFiles3 <- map(dataFiles2,separateLOCUSID)

for (i in seq_along(dataFiles3)) {
  dataFiles3[[i]]$chr <- NULL
}
```

Combine dataframes from all GAGA LGTs in the large list element ("dataFiles3") to one big data frame ("df")
```{r combine files}
df_combined<-rbindlist(dataFiles3,idcol = T, fill=TRUE) #the fill=TRUE option fills missing columns
head(df_combined)
```

## Add information from all dfast files to CDS-level df (df_combined)
```{r add dfast}
# For a single file
df.dfast<-read.csv("/home/j/j_rink02/sciebo/Master.LGTs/dfast/GAGA-0515.Scaffold8.1760778-1761241.fa/genome.gff_mod.gff",sep="\t",header=F)
df.new<-merge(df, df.dfast,by.x=c("locus.scaffold","CDS.start","CDS.end"),by.y=c("V1","V4","V5"),all.x=T,all.y=T)

df.new<-df.new %>%
  dplyr::relocate(".id", .before = "locus.scaffold") %>%
  dplyr::relocate("locus.position", .before = "CDS.start")

df.new$product<-gsub(".*product=(.*?)\\;.*","\\1",df.new$V9)
df.new$UniProtID<-gsub(".*UniprotKB:UniRef90_(.*?)\\;.*","\\1",df.new$V9)
df.new$dfastBacteriapredicted<-gsub(".*\\((.*?)\\).*","\\1",df.new$V9)

# For all files
gff2load<-"/home/j/j_rink02/sciebo/Master.LGTs/dfast/*/genome.gff_mod.gff"
gff.files.to.load<-Sys.glob(gff2load) #Sys.glob() returns all paths to the files based on a wildcard pattern in our file paths
head(gff.files.to.load)

gffs <- lapply(gff.files.to.load,read.csv,sep="\t",header=F) #lapply function to read all files and save them as dataframes
names(gffs)<-gsub("/home/j/j_rink02/sciebo/Master.LGTs/dfast/*/(.*?)\\.fa.*","\\1",gff.files.to.load)

new_dfs<-merge(dataFiles3,gffs,by.x=c("locus.scaffold","CDS.start","CDS.end"),by.y=c("seqid","start","end"),all.x=T,all.y=T)

for (i in seq_along(dataFiles3)) {
  dataFiles4<-merge(dataFiles3[[i]],gffs[[i]],by.x=c("locus.scaffold","CDS.start","CDS.end"),by.y=c("seqid","start","end"),all.x=T,all.y=T)
}
```





## Summarize CDS-level table (df_combined) together to obtain the LGT level.
```{r summary CDS level}
# For a single file
df.final<-df.new %>%  
  dplyr::group_by(.id, locus.scaffold, locus.position) %>% 
  dplyr::summarize(CDS.starts = paste0(CDS.start, collapse = ";"),CDS.ends = paste0(CDS.end, collapse = ";"),CDS.totalcounts = paste0(totalcount, collapse = ";"),CDS.products = paste0(product, collapse = ";"),CDS.UniProtIDs = paste0(UniProtID, collapse = ";"),dfast.donor = paste0(dfastBacteriapredicted, collapse = ";")) 

# To extract pid, q_cov_s_cov
gsub(".*?\\[(.*?)\\].*","\\1",df.dfast$V9)

# For all files


```


## Merge information for all LGTs together with the LGT candidate data table (497 candidates).
For every LGT candidate region, CDS starts, CDS ends and CDS total counts will be given in summarized columns.
```{r load lgts-combined table}
library(readxl)
LGTs.combined<-read_excel("/Users/Janina/Documents/MASTER/Masterarbeit/lgt/0_data/LGTs.combined.dfast.xlsx")
```

`r nrow(LGTs.combined)` CDS were predicted by dfast.





## Merge information for all CDS together with the CDS candidate data table (which also includes s_cov and q_cov.


Calculate the expression for the different life stages of GAGA-0515
```{r expression GAGA-0515}
#dataFiles[["GAGA-0515.Scaffold8.1760778-1761241"]]
i<-393
G515<-dataFiles[[i]]

# Add a column for the different life stages to calculate the reads separately for every life stage
G515$stage<-gsub(".*Cobs_(.*?)\\_.*","\\1",G515$e_id)

#Calculate the reads for every life stage
G515_df<-G515 %>%                               
  dplyr::group_by(chr,start,end,stage) %>%           
  dplyr::summarise_at(vars(ucount),         
               list(totalcountPerStage = sum)) 
  
# Calculate mean of samples
G515_mean<-G515 %>%
  dplyr::group_by(chr,start,end,stage) %>% 
  dplyr::summarise(avg=round(mean(ucount),1)) # with round() you can specify how many decimal places you want

G515.new<-merge(G515_df, G515_mean,by.x=c("chr","start","end","stage"),by.y=c("chr","start","end","stage"),all.x=T,all.y=T)

# Include the number of samples per stage
G515_sampleCount<-G515 %>%
  dplyr::count(stage) 

# Merge all information together into a final G-0515 expression table
G515.expression<-merge(G515.new, G515_sampleCount,by.x=c("stage"),by.y=c("stage"),all.x=T,all.y=T)

G515.expression$locusID<-names(dataFiles)[393]             # Add the name of the LGT by adding a column called "locusID" 
G515.expression<-G515.expression %>% 
  separate(locusID, sep="\\.", c(".id","locus.scaffold","locus.position")) %>% 
  dplyr::rename(CDS.start=start, CDS.end=end)# Separate the locusID column into .id, scaffold and CDS position

G515.expression<-G515.expression %>%
  dplyr::select(-chr)%>%
  dplyr::relocate("CDS.start","CDS.end","stage","totalcountPerStage","n","avg", .after = "locus.position")

G515.expression

# Save output as a table
write.table(G515.expression,"/home/j/j_rink02/sciebo/Master.LGTs/dfast/GAGA-0515.Scaffold8.1760778-1761241.fa/G515.expression.tsv",sep="\t",quote = F,row.names = F)
```







