---
title: "R Notebook"
output: html_notebook
---

```{r}
library(tidyr)
library(GenomicAlignments)
library(rtracklayer)
library(regioneR)
library(ggplot2)
library(cowplot)
```

```{r}
# define plotting function
plotCoverage <- function(bamfile, dfast.gff, lgt.bed, outfolder, id, expand) {
  # load data
  bam1<-readGAlignments(bamfile)      #load bam file
  gffCDS<-readGFFAsGRanges(dfast.gff) #load dfast annotation
  gffLGT<-toGRanges(lgt.bed)          #load lgt candidate bed
  
  
  # transform data for plotting
  gffLGT$type<-gffLGT$name        #assign LGT as type for the LGT bed file entries
  gr1<-GRangesList(gffCDS,gffLGT) #combine CDS and LGT info in one object
  bam1cov <- coverage(bam1)       #extract coverage info from bam file
  plottingRange <- GRanges(gffLGT@seqnames[1],IRanges(start(gffLGT)-expand,end(gffLGT)+expand)) #expand CDS range

  # extract the coverage info from the bed file  
  bam1cov.scf<-bam1cov[[plottingRange@seqnames[1]]]
  bam1covNum <- as.numeric(bam1cov.scf[ranges(plottingRange)]) #Uncompress the coverage

  # prepare a data frame with the coverage relative to the scf coordinates
  RNAseq<-data.frame(start=start(plottingRange)+seq(0,(length(bam1covNum)-1)),coverage=bam1covNum)
  
  #plot the CDS and LGT regions
  p.annotation <- ggplot(as.data.frame(gr1)) +
    geom_segment(aes(
      x = start,
      xend = end,
      col = type,
      y = type,
      yend = type
    ), lwd = 5) +
    theme_cowplot(font_size = 20) +
    theme(legend.position = "none",
          panel.spacing  = unit(c(0, 0, 0, 0), "cm")) +
    coord_cartesian(xlim = c(start(plottingRange), end(plottingRange))) +
    scale_color_manual(values = c("turquoise4", "grey80")) +
    ylab(NULL) +
    xlab(paste(id," ",gffLGT@seqnames,":",start(gffLGT),"-",end(gffLGT),sep = "")) +
    scale_x_continuous(expand = c(0, 0))
  
  #plot the coverage
  p.coverage <-
    ggplot(RNAseq) + geom_bar(
      stat = "identity",
      aes(x = start, y = coverage),
      fill = "skyblue4",
      col = "skyblue4",
      width = 1.1
    ) +
    coord_cartesian(xlim = c(start(plottingRange), end(plottingRange))) +
    theme_cowplot(font_size = 20) +
    xlab(NULL) +
    ylab(NULL) +
    scale_y_continuous(breaks = max(RNAseq$coverage)) +
    theme(
      plot.margin = unit(c(0, 0, -1, 0), "cm"),
      axis.title.x = element_blank(),
      axis.text.x = element_blank(),
      axis.ticks.x = element_blank(),
      axis.line.x = element_blank()
    )

  
#combine both plots
  p.full <-
    plot_grid(
      p.coverage,
      p.annotation,
      nrow = 2,
      align = "v",
      rel_heights = c(3 / 5, 2 / 5)
    )

  #print plot and save
  p.full
  ggsave(paste(outfolder,id,"-",gffLGT@seqnames,"-",start(gffLGT),"-",end(gffLGT),".uniqueReads.coverage.pdf",sep=""),width=8,height=4)
}
```

```{r}
# prepare vectors for ids, bamfiles, dfasts and beds for all the LGT canidates you want to plot. This can come from a tsv file or an excel file or whatever.

# Upload a file with the file names
LGT_df<-read.csv2("/Users/Janina/Documents/MASTER/Masterarbeit/lgt/0_data/Tables/LGT_filenames.csv")

# Vector for the ids
ids<-c(LGT_df$id)

# Vector for the bam files
bamfiles<-c(LGT_df$bam)

# Vector for the dfast gff files
dfasts<-c(LGT_df$gff)

# Vector for the bed files
beds<-c(LGT_df$bed)
```

```{r}
#loop over all entries in id, bamfiles, dfasts, and beds to create one pdf per lgt candidate region
for (i in 1:length(LGT_df$id)){
plotCoverage(
  id=LGT_df$id[i],
  bamfile=LGT_df$bam[i],
  dfast.gff=LGT_df$gff[i],
  lgt.bed=LGT_df$bed[i],
  outfolder="/Users/Janina/sciebo/Master.LGTs/RNAseqmapping/RNAseq_plots/UniqueReads/",
  expand=100  #how far should the plot expanded around the LGT
)
}
```

## Plot the unique read counts of the reannotated LGT candidates
```{r}
Unique_reannotated_LGTs<-read.csv2("/Users/Janina/Documents/MASTER/Masterarbeit/lgt/0_data/Tables/LGT_filenames_reannotated.csv")

# Vector for the ids
ids<-c(Unique_reannotated_LGTs$id)
# Vector for the bam files
bamfiles<-c(Unique_reannotated_LGTs$bam)
# Vector for the dfast gff files
dfasts<-c(Unique_reannotated_LGTs$gff)
# Vector for the bed files
beds<-c(Unique_reannotated_LGTs$bed)

for (i in 1:length(Unique_reannotated_LGTs$id)){
plotCoverage(
  id=Unique_reannotated_LGTs$id[i],
  bamfile=Unique_reannotated_LGTs$bam[i],
  dfast.gff=Unique_reannotated_LGTs$gff[i],
  lgt.bed=Unique_reannotated_LGTs$bed[i],
  outfolder="/Users/Janina/sciebo/Master.LGTs/RNAseqmapping/RNAseq_plots/UniqueReads_reannotated/",
  expand=200  #how far should the plot expanded around the LGT
)
}
```

```{r debugging}
ids<-c("GAGA-0099")
bamfiles<-c("/Users/Janina/sciebo/Master.LGTs/RNAseqmapping/reannotation.dfast_RNAseq_mapping/GAGA-0099/GAGA-0099.Scaffold15.2285014-2297549.1000.out/uniquely_mapped.LGTregion.sorted.bam")
dfasts<-c("/Users/Janina/sciebo/Master.LGTs/reannotation.dfast/GAGA-0099.Scaffold15.2285014-2297549.1000.out/genome_reannotate_mod_intersect_modified.gff")
beds<-c("/Users/Janina/sciebo/Master.LGTs/dfast/GAGA-0099.Scaffold15.2285014-2297549.fa/genome.gff.bed")

# Debug code from function
bam1<-readGAlignments("/Users/Janina/sciebo/Master.LGTs/RNAseqmapping/reannotation.dfast_RNAseq_mapping/GAGA-0099/GAGA-0099.Scaffold15.2285014-2297549.1000.out/uniquely_mapped.LGTregion.sorted.bam")      #load bam file
gffCDS<-readGFFAsGRanges("/Users/Janina/sciebo/Master.LGTs/reannotation.dfast/GAGA-0099.Scaffold15.2285014-2297549.1000.out/genome_reannotate_mod_intersect.gff") #load dfast annotation
gffLGT<-toGRanges("/Users/Janina/sciebo/Master.LGTs/dfast/GAGA-0099.Scaffold15.2285014-2297549.fa/genome.gff.bed")  

for (i in 1:length(ids)){
plotCoverage(
  id=ids[i],
  bamfile=bamfiles[i],
  dfast.gff=dfasts[i],
  lgt.bed=beds[i],
  outfolder="/Users/Janina/sciebo/Master.LGTs/RNAseqmapping/RNAseq_plots/UniqueReads_reannotated/",
  expand=100  #how far should the plot expanded around the LGT
)
}
```



